### PR TITLE
Better mocking for Core Unit Tests

### PR DESCRIPTION
**Pull request by ryanwilson**
This includes NSNotificationCenter and NSUserDefaults to prevent
flaky tests and unit tests interfering with each other.

### CODE REVIEW

File: Example/Core/Tests/FIRAnalyticsConfigurationTest.m [hunk 1]
```diff
@@ -44,49 +50,69 @@ - (void)testSharedInstance {
...
-  OCMVerify([self.notificationCenterMock postNotificationName:notificationName
-                                                       object:config
-                                                     userInfo:@{
-                                                       notificationName : @2601
-                                                     }]);
+  [self expectNotificationForObserver:self.observerMock
+                     notificationName:notificationName
+                               object:config
+                             userInfo:@{
+                               notificationName : @2601
```
Comment (paulb777) [inline Example/Core/Tests/FIRAnalyticsConfigurationTest.m]: why 2601?

File: Example/Core/Tests/FIRAnalyticsConfigurationTest.m [hunk 2]
```diff
@@ -44,49 +50,69 @@ - (void)testSharedInstance {
...
-  OCMVerify([self.notificationCenterMock postNotificationName:notificationName
-                                                       object:config
-                                                     userInfo:@{
-                                                       notificationName : @1000
-                                                     }]);
+  [self expectNotificationForObserver:self.observerMock
+                     notificationName:notificationName
+                               object:config
+                             userInfo:@{
+                               notificationName : @1000
```
Comment (paulb777) [inline Example/Core/Tests/FIRAnalyticsConfigurationTest.m]: why 1000?

File: Firebase/Core/FIRLogger.m [hunk 3]
```diff
@@ -190,7 +197,12 @@ void FIRSetLoggerLevel(FIRLoggerLevel loggerLevel) {
 #ifdef DEBUG
 void FIRResetLogger() {
   sFIRLoggerOnceToken = 0;
-  [[NSUserDefaults standardUserDefaults] removeObjectForKey:kFIRPersistedDebugModeKey];
+  [sFIRLoggerUserDefaults removeObjectForKey:kFIRPersistedDebugModeKey];
+  sFIRLoggerUserDefaults = nil;
+}
+
+void FIRSetLoggerUserDefaults(NSUserDefaults *defaults) {
```
Comment (paulb777) [inline Firebase/Core/FIRLogger.m]: I think this won't work if DEBUG isn't on.

File: Example/Core/Tests/FIRAnalyticsConfigurationTest.m (hunk 1 above)
Comment (ryanwilson) [reply Example/Core/Tests/FIRAnalyticsConfigurationTest.m]: These were randomly picked previously and the number is passed in below. It made more sense before when the method calling it was above :) I'll pull it out into a variable and use it here as well as the call below.

File: Example/Core/Tests/FIRAnalyticsConfigurationTest.m (hunk 2 above)
Comment (ryanwilson) [reply Example/Core/Tests/FIRAnalyticsConfigurationTest.m]: Same as above.

File: Firebase/Core/FIRLogger.m (hunk 3 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: I was going to only use it for unit testing at the moment, similar to `FIRResetLogger()`. If we were to move to a separate user defaults system for Core or all of Firebase (which we probably should? we can discuss that separately) then we could move it out of the `#ifdef DEBUG` check.

Does that make sense, or is there a different concern?

**Pull request merged**